### ISSUE TITLE
Acessing a possibly null property is not detected

### ISSUE THREAD

**Issue by ihsoft** [google/closure-compiler#1146]
```javascript
// ==ClosureCompiler==
// @compilation_level ADVANCED_OPTIMIZATIONS
// @output_file_name default.js
// ==/ClosureCompiler==

/** @struct @constructor */
var A = function A() {
  this.t = 1;
};

/** @struct @constructor */
var B = function B() {
  /** @type {?A} */
  this.a = null;
};

/** @return {!number} */
B.prototype.f1 = function() {
  return this.a.t;  // <== OK, but expected WARNING
};

var b = new B();
alert(b.f1());
```
Compiled code is:
```javascript
alert((new function(){this.a=null}).a.t);
```
For some reason in statement `return this.a.t` compiler assumes `B.a` to be `{!A}` and doesn't show a "type mismatch" warning.

**Comment by Dominator008** [comment]
The Compiler's new type inference (NTI) actually checks for it, but the warning is hidden behind a DiagnosticGroup `newCheckTypesAllChecks` because many internal Google projects have nullable dereferences.

To see the warning, you can use the following compilation command from the command line:
```
java -jar compiler.jar -W VERBOSE --new_type_inf --jscomp_warning=newCheckTypesAllChecks your_file.js
```

We should consider giving the DiagnosticGroup a more descriptive name and document it better.

**Comment by concavelenz** [comment]
There is also the NullDeref conformance check that can be used with the OTI

**Comment by chrismiddleton** [comment]
I'm not sure if this is the same issue, but I noticed in my own program that a possibly null return value was not generating an error when used as a non-null value. Here is a simple program which shows the problem:

```
// ==ClosureCompiler==
// @compilation_level ADVANCED_OPTIMIZATIONS
// @output_file_name default.js
// ==/ClosureCompiler==

/**
 * @constructor
**/
function Bar () {
}

Bar.prototype.bark = function () {
  alert("WOOF!");
};

/**
 * @return {Bar!}
**/
Bar.prototype.clone = function () {
  return new Bar();
};

/**
 * @constructor
**/
function Foo () {
  this._bar = Math.random() < 0.5 ? new Bar() : null;
}

/**
 * @return {Bar?}
**/
Foo.prototype.getBar = function () {
  return this._bar;
};

// !! This will throw a TypeError half the time, but gives no error or warning when compiling
var bar = (new Foo()).getBar().clone();
bar.bark();
```

https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250A%252F**%250A%2520*%2520%2540constructor%250A**%252F%250Afunction%2520Bar%2520%28%29%2520%257B%250A%257D%250A%250ABar.prototype.bark%2520%253D%2520function%2520%28%29%2520%257B%250A%2520%2520alert%28%2522WOOF!%2522%29%253B%250A%257D%253B%250A%250A%252F**%250A%2520*%2520%2540return%2520%257BBar!%257D%250A**%252F%250ABar.prototype.clone%2520%253D%2520function%2520%28%29%2520%257B%250A%2520%2520return%2520new%2520Bar%28%29%253B%250A%257D%253B%250A%250A%252F**%250A%2520*%2520%2540constructor%250A**%252F%250Afunction%2520Foo%2520%28%29%2520%257B%250A%2520%2520this._bar%2520%253D%2520Math.random%28%29%2520%253C%25200.5%2520%253F%2520new%2520Bar%28%29%2520%253A%2520null%253B%250A%257D%250A%250A%252F**%250A%2520*%2520%2540return%2520%257BBar%253F%257D%250A**%252F%250AFoo.prototype.getBar%2520%253D%2520function%2520%28%29%2520%257B%250A%2520%2520return%2520this._bar%253B%250A%257D%253B%250A%250Avar%2520bar%2520%253D%2520%28new%2520Foo%28%29%29.getBar%28%29.clone%28%29%253B%250Abar.bark%28%29%253B%250A%250A

**Comment by chrismiddleton** [comment]
To add to the weirdness, when you compile the above with ADVANCED_OPTIMIZATIONS on, but change the name "clone" to something that can be minified, like "clome", it compiles to simply `alert("WOOF!")`, which actually changes the behavior of the program (by removing the Math.random portion). Presumably the same bug causing it to think that getBar() returns a !Bar instead of a ?Bar makes it think it can simplify the entire program to a single function call. This is actually more troubling than the lack of warning/error.

Appspot example: 

https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250A%252F**%250A%2520*%2520%2540constructor%250A**%252F%250Afunction%2520Bar%2520%28%29%2520%257B%250A%257D%250A%250ABar.prototype.bark%2520%253D%2520function%2520%28%29%2520%257B%250A%2520%2520alert%28%2522WOOF!%2522%29%253B%250A%257D%253B%250A%250A%252F**%250A%2520*%2520%2540return%2520%257BBar!%257D%250A**%252F%250ABar.prototype.clome%2520%253D%2520function%2520%28%29%2520%257B%250A%2520%2520return%2520new%2520Bar%28%29%253B%250A%257D%253B%250A%250A%252F**%250A%2520*%2520%2540constructor%250A**%252F%250Afunction%2520Foo%2520%28%29%2520%257B%250A%2520%2520this._bar%2520%253D%2520Math.random%28%29%2520%253C%25200.5%2520%253F%2520new%2520Bar%28%29%2520%253A%2520null%253B%250A%257D%250A%250A%252F**%250A%2520*%2520%2540return%2520%257BBar%253F%257D%250A**%252F%250AFoo.prototype.getBar%2520%253D%2520function%2520%28%29%2520%257B%250A%2520%2520return%2520this._bar%253B%250A%257D%253B%250A%250Avar%2520bar%2520%253D%2520%28new%2520Foo%28%29%29.getBar%28%29.clome%28%29%253B%250Abar.bark%28%29%253B%250A%250A

**Comment by brad4d** [comment]
@chrismiddleton, it is the same issue.
Here's an example in the debugger of the option turned on to detect the problem.
https://goo.gl/lYLViT

The second behavior is actually as expected. 
Since the nullability check is turned off the compiler assumes there must actually be a Bar object on which to call bark(). From there the compiler can tell that the end result is just to call alert() and is able to abstract everything else away.

It only keeps everything when the method is named 'clone', because 'clone' is one of the standard exported names it's not allowed to rename or remove.

**Comment by chrismiddleton** [comment]
> The second behavior is actually as expected. 

Perhaps it is as expected given the current logic, but I don't see how it's expected from a user point of view. My assumption has generally been "turning off warnings disables the warning", not "turning on a warning prevents the code's behavior from being changed". I would think that if nullability checks are off, it shouldn't be making an assumption that something is non-null. If it's not checking nullability, how can it rightly make that assumption?

I'm trying to get the new type inference working in my code to see if it's not an issue there,.

**Comment by dimvar** [comment]
I looked into this and I don't think it's purely a documentation issue. It would be good to address the bug that changes the semantics of the code. Here is the example, slightly simplified, that also has the bug in advanced mode.

    /** @constructor */
    function Bar () {}
    
    Bar.prototype.bark = function () {
      alert("WOOF!");
    };
    
    /** @constructor */
    function Foo () {
      /** @type {?Bar} */
      this._bar = Math.random() < 0.5 ? new Bar() : null;
    }
    
    (new Foo())._bar.bark();

The wrong rewriting happens in method devirtualization. The method is turned to a function that doesn't use its argument. I won't have time to look into this more now, but leaving the comment for whoever wants to investigate at some point.

    // parseInputs yields:
    // ************************************
    function Bar() {
    }
    Bar.prototype.bark = function() {
      alert("WOOF!");
    };
    function Foo() {
      this._bar = Math.random() < .5 ? new Bar : null;
    }
    (new Foo)._bar.bark();
    
    
    // inlineVariables yields:
    // ************************************
    function Bar() {
    }
    Bar.prototype.bark = function() {
      alert("WOOF!");
    };
    (new function Foo() {
      this._bar = Math.random() < .5 ? new Bar : null;
    })._bar.bark();
    
    
    // devirtualizePrototypeMethods yields:
    // ************************************
    function Bar() {
    }
    var JSCompiler_StaticMethods_bark = function(JSCompiler_StaticMethods_bark$self) {
      alert("WOOF!");
    };
    JSCompiler_StaticMethods_bark((new function Foo() {
      this._bar = Math.random() < .5 ? new Bar : null;
    })._bar);

**Comment by dimvar** [comment]
I compiled the code with advanced mode and the flags: --print_source_after_each_pass, --pretty_print,
--generate_pseudo_names.

**Comment by chrismiddleton** [comment]
@dimvar Thanks for taking a look into this so quickly. Also cool to see some of the debugging.
